src/Microsoft.ML.Tokenizers/Model/SentencePieceUnigramModel.cs

```diff
@@ -62,8 +62,24 @@ public SentencePieceUnigramModel(ModelProto modelProto, bool addBos, bool addEos
 
             _trie = new DoubleArrayTrie(_vocab);
 
-            _vocabReverse[BeginningOfSentenceId] = (BeginningOfSentenceToken, 0f, 0);
-            _vocabReverse[EndOfSentenceId] = (EndOfSentenceToken, 0f, 0);
+
+            Debug.Assert(modelProto.TrainerSpec.UnkId >= 0);
+            Debug.Assert(modelProto.TrainerSpec.BosId >= 0);
+            Debug.Assert(modelProto.TrainerSpec.EosId >= 0);
+
+            _vocab[modelProto.TrainerSpec.UnkPiece] = modelProto.TrainerSpec.UnkId;
```

Comment:
Was this present in the original tokenizer or is it special to our port?  Asking because I don't understand why this isn't handled via `modelProto.Pieces`. Are we certain we only need these 3 special cases?  Comment might help.

Reply:
This is special to our port to ensure adding these special tokens to the vocabs for easier lookup. Adding these here are not changing any behavior more than allowing the vocabulary to map these tokens which help us in some operations like decoding for example. 

I'll add some detailed comment here. Thanks!

Reply:
Can you know for sure that `modelProto.Pieces.Count` used when allocating `_vocabReverse` is greater than these IDs?  I guess so since these IDs come from the modelProto too.  Just seems odd to me that we step through all the `modelProto.Pieces.Count` above, but then we come back and overwrite some of the IDs like this.

Reply:
> Just seems odd to me that we step through all the modelProto.Pieces.Count above, but then we come back and overwrite some of the IDs like this.

This is done this way to avoid adding these special tokens to the `trie` as these shouldn't be part of it. The native code is also not adding such tokens when enumerating the `modelProto.Pieces`. It is only our addition is we add these to the vocab after we are done building the trie for easier mapping internally.